src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs

```diff
@@ -39,6 +39,38 @@ public CertificateCallbackMapper(Func<HttpRequestMessage, X509Certificate2, X509
         }
 
         public static async ValueTask<(Socket, Stream)> ConnectAsync(string host, int port, CancellationToken cancellationToken)
+        {
+            CancellationTokenSource cancelIPv6 = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            Task<(Socket, Stream)> tryConnectAsyncIPv6 = ConnectAsyncInternal(host, port, AddressFamily.InterNetworkV6, cancelIPv6.Token);
+
+            if (await Task.WhenAny(tryConnectAsyncIPv6, Task.Delay(200)).ConfigureAwait(false) == tryConnectAsyncIPv6 && tryConnectAsyncIPv6.IsCompletedSuccessfully)
+            {
+                return await tryConnectAsyncIPv6.ConfigureAwait(false);
+            }
+
+            CancellationTokenSource cancelIPv4 = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            Task<(Socket, Stream)> tryConnectAsyncIPv4 = ConnectAsyncInternal(host, port, AddressFamily.InterNetwork, cancelIPv4.Token);
```

Comment:
Also, each of these ConnectAsyncInternal calls is going to do its own Dns lookup.  Two concerns with that:
1. There's overhead to each.
2. What if the results come back different for the two requests?

Seems like the main method should actually do the Dns request, and then either order the results as is outlined in the RFC and have the two connect methods iterate through that in a thread-safe manner, or partition it into two arrays (one for IPv6, one for IPv4) that are then handed off to these connect methods.

Reply:
I agree, duplicating the DNS lookup isn't the way to go. Keeping control over that process would also make it easier to manage the lifetimes of the sockets involved, which I think would help resolve some of the other issues you've brought up.

My first thought is that I should add something along the lines of `DualSocketMultipleConnectAsync` to `SocketsHttpHandler` (with async/await instead of APM). Does that sound like the right approach to you?